### PR TITLE
Add validation to Insert statement parsing to match the column names with supplied values

### LINKED ISSUE #3069: Insert statement does not fully validate the column names with supplied values
For instance, 

We can successfully execute a statement like: 
```
INSERT INTO T(a, b, c) VALUES (1, 2, 3, 4, 5)
```
which should actually fail. 

Moreover, statement like: 
```
INSERT INTO T(a, b, c) VALUES (1)
```
should pass only if `b` and `c` are nullable or have defualt value, else should fail.

### PR DESCRIPTION
**Pull request by g31pranjal**
fixes #3069. This PR also provide a short-term solution to #3068 

In essence, it does the following things: 
1. Throws an exception, if a column is declared in the `INSERT` statement but its value is not provided.
2. Throws an exception, if the number of columns declared in the `INSERT` statement is not equal to the number of values provided for a row
3. Throws an exception, if a not-null column is not declared in the `INSERT` statement or its value is not provided
4. Restricts declaring types (except `ARRAY`) as `NOT NULL` in the DDL since this feature is not fully supported throughout. 
5. Removes providing system default value in any scenario. {Since we do not formally support `DEFAULT`}

### CODE REVIEW

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 1]
```diff
@@ -124,18 +125,14 @@ private void insertArraysViaQuerySimpleStatement() throws SQLException {
...
-        insertQuery("INSERT INTO T (pk) VALUES (4)");
+        Assert.assertThrows(SQLException.class, () -> insertQuery("INSERT INTO T (pk) VALUES (4)"));
     }
 
     @Test
     void testInsertArraysViaQueryPreparedStatement() throws SQLException {
         final var statement = "INSERT INTO T (pk, boolean_null, boolean_not_null, integer_null, integer_not_null, " +
                 "bigint_null, bigint_not_null, float_null, float_not_null, double_null, double_not_null, string_null, " +
-                "string_not_null, bytes_null, bytes_not_null, struct_null, struct_not_null) VALUES (?pk, ?boolean_null, " +
+                "string_not_null, bytes_null, bytes_not_null, structure_null, structure_not_null) VALUES (?pk, ?boolean_null, " +
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:135]: You probably want to revert back to `struct_null` instead of `structure_null` since the type name is `STRUCT`

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 2]
```diff
@@ -124,18 +125,14 @@ private void insertArraysViaQuerySimpleStatement() throws SQLException {
                         "[x'31', x'32'], null, " +
                         "[(11, '11'), (22, '22')], null " +
                         ")"));
-        // This should not go through since non-null array is not initialized, However, that is being currently getting
-        // initialized by an empty array (no elements in unwrapped REPEATED field in proto).
-        // This phenomenon is also true with basic types, i.e., non-nullability is not enforced and is tracked by
-        // TODO (Add support + tests for column nullable/not null)
-        insertQuery("INSERT INTO T (pk) VALUES (4)");
+        Assert.assertThrows(SQLException.class, () -> insertQuery("INSERT INTO T (pk) VALUES (4)"));
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:128]: Can you use `RelationalAssertions.assertThrowsSqlException` that allows you to check for the error message and SQLState?

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 3]
```diff
@@ -216,6 +213,116 @@ void testInsertArraysViaQueryPreparedStatement() throws SQLException {
...
+
+    @Test
+    void testNonNullViolation() throws SQLException {
+        try (final var conn = DriverManager.getConnection(database.getConnectionUri().toString())) {
+            conn.setSchema(database.getSchemaName());
+            conn.setAutoCommit(true);
+            try (final var ps = ((RelationalPreparedStatement) conn.prepareStatement("INSERT INTO T (pk) VALUES (?pk)"))) {
+                ps.setInt("pk", 1);
+                final var actualMessage = Assertions.assertThrows(SQLException.class, ps::executeUpdate).getMessage();
+                System.out.println(actualMessage);
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:239]: remove print

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 4]
```diff
@@ -216,6 +213,116 @@ void testInsertArraysViaQueryPreparedStatement() throws SQLException {
...
+    @Test
+    void testMoreParametersThanColumns() throws SQLException {
+        try (final var conn = DriverManager.getConnection(database.getConnectionUri().toString())) {
+            conn.setSchema(database.getSchemaName());
+            conn.setAutoCommit(true);
+            try (final var ps = ((RelationalPreparedStatement) conn.prepareStatement("INSERT INTO T (pk) VALUES (?pk, ?boolean_null, ?boolean_not_null)"))) {
+                ps.setInt("pk", 1);
+                ps.setArray("boolean_null", EmbeddedRelationalArray.newBuilder().addAll(true, false).build());
+                ps.setArray("boolean_not_null", EmbeddedRelationalArray.newBuilder().addAll(true, false).build());
+                final var actualMessage = Assertions.assertThrows(SQLException.class, ps::executeUpdate).getMessage();
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:225]: Use `RelationalAssertions.assertThrowSqlException`

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 5]
```diff
@@ -216,6 +213,116 @@ void testInsertArraysViaQueryPreparedStatement() throws SQLException {
...
+    }
+
+    @Test
+    void testNonNullViolation() throws SQLException {
+        try (final var conn = DriverManager.getConnection(database.getConnectionUri().toString())) {
+            conn.setSchema(database.getSchemaName());
+            conn.setAutoCommit(true);
+            try (final var ps = ((RelationalPreparedStatement) conn.prepareStatement("INSERT INTO T (pk) VALUES (?pk)"))) {
+                ps.setInt("pk", 1);
+                final var actualMessage = Assertions.assertThrows(SQLException.class, ps::executeUpdate).getMessage();
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:238]: use `RelationalAssertions.assertThrowSqlException`

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 6]
```diff
@@ -216,6 +213,116 @@ void testInsertArraysViaQueryPreparedStatement() throws SQLException {
...
+    }
+
+    @Test
+    void testNullColumnDoesNotRequireAValueIfNotGiven() throws SQLException {
+        try (final var conn = DriverManager.getConnection(database.getConnectionUri().toString())) {
+            conn.setSchema(database.getSchemaName());
+            conn.setAutoCommit(true);
+            try (final var ps = ((RelationalPreparedStatement) conn.prepareStatement("INSERT INTO T (pk, boolean_not_null, " +
+                    "integer_not_null, bigint_not_null, float_not_null, double_not_null, string_not_null, bytes_not_null, " +
+                    "structure_not_null, boolean_null, integer_null) VALUES (?pk, ?boolean_not_null, ?integer_not_null, " +
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:288]: This should fail. If we specify the column in the `insert into T(<columns>), then it must be required in the list that follows.

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 7]
```diff
@@ -216,6 +213,116 @@ void testInsertArraysViaQueryPreparedStatement() throws SQLException {
...
+                ps.setArray("string_not_null", EmbeddedRelationalArray.newBuilder().addAll("11", "22").build());
+                ps.setArray("bytes_not_null", EmbeddedRelationalArray.newBuilder().addAll(new byte[]{49}, new byte[]{50}).build());
+                ps.setArray("struct_not_null", EmbeddedRelationalArray.newBuilder().addAll(EmbeddedRelationalStruct.newBuilder().addInt("a", 11).addString("b", "11").build(), EmbeddedRelationalStruct.newBuilder().addInt("a", 22).addString("b", "22").build()).build());
+                assertEquals(1, ps.executeUpdate());
+            }
+        }
+    }
+
+    @Test
+    void testNonNullColumnRequiresValue() throws SQLException {
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:305]: This is misleading. The column is not `NOT NULL` since this PR does away with this concept. A column declared in `INSERT INTO T(<columns>)` must be in the `VALUES` clause though.

**Review by arnaud-lacurie — CHANGES_REQUESTED** [review]
Can you also fix up the title of the PR and add a more comprehensive description?

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 2 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:128]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 3 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:239]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 4 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:225]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 5 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:238]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 1 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:135]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 6 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:288]: Okay, I miswrote the logic for this. Fixed.

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 7 above)
Comment (g31pranjal) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:305]: This is fixed as well!

File: yaml-tests/src/test/resources/inserts-updates-deletes.yamsql [hunk 8]
```diff
@@ -105,4 +105,14 @@ test_block:
       - result: [{ B1: 10, B2: 22, B3: { 5, 41 } },
                  { B1: 20, B2: 22, B3: { 6, 51 } },
                  { B1: 30, B2: 22, B3: { 7, 61 } } ]
+    -
+      # Case where not all values are provided of A. Still works, since the columns for which no values are provided can be nullable.
```
[the remark this answers is not in the document]
Comment (g31pranjal) [reply yaml-tests/src/test/resources/inserts-updates-deletes.yamsql:109]: Fixed this! Added the test case that you mentioned.

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java (hunk 2 above)
Comment (arnaud-lacurie) [reply fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java]: Can you check the SQLState as well? 
`RelationalAssertions.assertThrowsSqlException` has a method `withErrorCode` or something like this that you can use.

File: yaml-tests/src/test/resources/inserts-updates-deletes.yamsql [hunk 9]
```diff
@@ -109,10 +109,19 @@ test_block:
...
       - supported_version: !current_version
-      - count: 1
+      - error: "42601"
     -
       # Case when the number of values is more than the number of columns specified.
       - query: insert into A(A1, A2, A3) values (5, 6, 7, 8, 9);
       - supported_version: !current_version
       - error: "42601"
+    -
+      # Case when a nullable column's value is not provided
```
Comment (arnaud-lacurie) [inline yaml-tests/src/test/resources/inserts-updates-deletes.yamsql:119]: Just to cover that, could you add a test where the not provided column is the first one and another case where it's the last one?

Comment (g31pranjal) [reply yaml-tests/src/test/resources/inserts-updates-deletes.yamsql:119]: done!

File: fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java [hunk 10]
```diff
@@ -123,12 +124,9 @@ private void insertArraysViaQuerySimpleStatement() throws SQLException {
                         "['11', '22'], null, " +
                         "[x'31', x'32'], null, " +
                         "[(11, '11'), (22, '22')], null " +
-                        ")"));
-        // This should not go through since non-null array is not initialized, However, that is being currently getting
-        // initialized by an empty array (no elements in unwrapped REPEATED field in proto).
-        // This phenomenon is also true with basic types, i.e., non-nullability is not enforced and is tracked by
-        // TODO (Add support + tests for column nullable/not null)
-        insertQuery("INSERT INTO T (pk) VALUES (4)");
+                        ")")).hasErrorCode(ErrorCode.INTERNAL_ERROR);
```
Comment (g31pranjal) [inline fdb-relational-core/src/test/java/com/apple/foundationdb/relational/recordlayer/RelationalArrayTest.java:127]: This is throwing INTERNAL_ERROR currently because the check for non-nullability is currently not happening during the record construction. It directly happens during execution which throws verifyException. Opened an issue about it https://github.com/FoundationDB/fdb-record-layer/issues/3133
